File: backend/scripts/migr.mjs
/**
 * Apply SQL files from supabase/migrations in timestamp order.
 *
 * Setup:
 *   1. In Supabase Dashboard → Project Settings → Database, copy the
 *      "URI" connection string (Direct connection or Session pooler).
 *   2. Add to backend/.env:
 *        DATABASE_URL=postgresql://postgres.[ref]:YOUR_PASSWORD@...
 *
 * Usage:
 *   node scripts/migr.mjs
 *   node scripts/migr.mjs --from 20250404000000
 *
 * --from <14-digit timestamp>  Skip migrations whose filename timestamp is
 *   strictly before this value (useful if the DB already has earlier migrations).
 */

import pg from 'pg';
import { config } from 'dotenv';
import { readFile, readdir } from 'fs/promises';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
config({ path: join(__dirname, '..', '.env') });

const MIGRATIONS_DIR = join(__dirname, '..', '..', 'supabase', 'migrations');

function parseArgs() {
  const argv = process.argv.slice(2);
  let fromTs = null;
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--from' && argv[i + 1]) {
      fromTs = argv[i + 1].replace(/\D/g, '').slice(0, 14);
      if (fromTs.length !== 14) {
        console.error('--from expects a 14-digit migration timestamp, e.g. 20250404000000');
        process.exit(1);
      }
      break;
    }
    if (argv[i].startsWith('--from=')) {
      fromTs = argv[i].split('=')[1].replace(/\D/g, '').slice(0, 14);
      if (fromTs.length !== 14) {
        console.error('--from expects a 14-digit migration timestamp');
        process.exit(1);
      }
      break;
    }
  }
  return { fromTs };
}

function fileTimestamp(filename) {
  const m = filename.match(/^(\d{14})/);
  return m ? m[1] : '00000000000000';
}

async function ensureMigrationsTable(client) {
  await client.query(`
    create table if not exists public.schema_migrations (
      filename text primary key,
      applied_at timestamptz not null default now()
    );
  `);
}

async function main() {
  const { fromTs } = parseArgs();
  const url = process.env.DATABASE_URL;
  if (!url) {
    console.error(
      'Missing DATABASE_URL in backend/.env — copy the Postgres URI from Supabase → Settings → Database.'
    );
    process.exit(1);
  }

  let files;
  try {
    files = (await readdir(MIGRATIONS_DIR))
      .filter((f) => f.endsWith('.sql'))
      .sort();
  } catch (e) {
    console.error('Cannot read migrations folder:', MIGRATIONS_DIR, e.message);
    process.exit(1);
  }

  if (files.length === 0) {
    console.log('No .sql files in', MIGRATIONS_DIR);
    return;
  }

  const client = new pg.Client({
    connectionString: url,
    ssl: url.includes('localhost') ? false : { rejectUnauthorized: false },
  });

  await client.connect();
  try {
    await ensureMigrationsTable(client);

    for (const name of files) {
      if (fromTs && fileTimestamp(name) < fromTs) {
        console.log('[skip --from]', name);
        continue;
      }

      const { rows } = await client.query(
        'select 1 from public.schema_migrations where filename = $1',
        [name]
      );
      if (rows.length > 0) {
        console.log('[already applied]', name);
        continue;
      }

      const path = join(MIGRATIONS_DIR, name);
      const sql = await readFile(path, 'utf8');
      console.log('[running]', name);
      await client.query('BEGIN');
      try {
        await client.query(sql);
        await client.query('insert into public.schema_migrations (filename) values ($1)', [
          name,
        ]);
        await client.query('COMMIT');
        console.log('[ok]', name);
      } catch (e) {
        await client.query('ROLLBACK');
        console.error('[failed]', name, e.message);
        process.exit(1);
      }
    }
    console.log('Done.');
  } finally {
    await client.end();
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: backend/scripts/migr.py
"""
Apply SQL files from supabase/migrations in timestamp order.

Setup:
  1. In Supabase Dashboard → Project Settings → Database, copy the
     "URI" connection string (Direct connection or Session pooler).
  2. Add to backend/.env:
       DATABASE_URL=postgresql://postgres.[ref]:YOUR_PASSWORD@...

Usage:
  python -m scripts.migr
  python -m scripts.migr --from 20250404000000
"""

from __future__ import annotations

import argparse
import os
import re
import sys
from pathlib import Path

import psycopg
from dotenv import load_dotenv

BACKEND_DIR = Path(__file__).resolve().parent.parent
MIGRATIONS_DIR = BACKEND_DIR.parent / "supabase" / "migrations"

load_dotenv(BACKEND_DIR / ".env")


def file_timestamp(filename: str) -> str:
    m = re.match(r"^(\d{14})", filename)
    return m.group(1) if m else "00000000000000"


def ensure_migrations_table(conn: psycopg.Connection) -> None:
    conn.execute(
        """
        create table if not exists public.schema_migrations (
          filename text primary key,
          applied_at timestamptz not null default now()
        );
        """
    )


def main() -> None:
    parser = argparse.ArgumentParser(description="Apply Supabase SQL migrations")
    parser.add_argument(
        "--from",
        dest="from_ts",
        default=None,
        help="Skip migrations with timestamp strictly before this 14-digit value",
    )
    args = parser.parse_args()
    from_ts = None
    if args.from_ts:
        from_ts = re.sub(r"\D", "", args.from_ts)[:14]
        if len(from_ts) != 14:
            print(
                "--from expects a 14-digit migration timestamp, e.g. 20250404000000",
                file=sys.stderr,
            )
            sys.exit(1)

    url = os.getenv("DATABASE_URL")
    if not url:
        print(
            "Missing DATABASE_URL in backend/.env — copy the Postgres URI from "
            "Supabase → Settings → Database.",
            file=sys.stderr,
        )
        sys.exit(1)

    try:
        files = sorted(f.name for f in MIGRATIONS_DIR.iterdir() if f.suffix == ".sql")
    except OSError as e:
        print("Cannot read migrations folder:", MIGRATIONS_DIR, e, file=sys.stderr)
        sys.exit(1)

    if not files:
        print("No .sql files in", MIGRATIONS_DIR)
        return

    connect_kwargs: dict = {"conninfo": url}
    if "localhost" not in url and "127.0.0.1" not in url:
        connect_kwargs["sslmode"] = "require"

    with psycopg.connect(**connect_kwargs) as conn:
        conn.autocommit = False
        ensure_migrations_table(conn)
        conn.commit()

        for name in files:
            if from_ts and file_timestamp(name) < from_ts:
                print("[skip --from]", name)
                continue

            row = conn.execute(
                "select 1 from public.schema_migrations where filename = %s",
                (name,),
            ).fetchone()
            if row:
                print("[already applied]", name)
                continue

            sql = (MIGRATIONS_DIR / name).read_text(encoding="utf-8")
            print("[running]", name)
            try:
                conn.execute(sql)
                conn.execute(
                    "insert into public.schema_migrations (filename) values (%s)",
                    (name,),
                )
                conn.commit()
                print("[ok]", name)
            except Exception as e:
                conn.rollback()
                print("[failed]", name, e, file=sys.stderr)
                sys.exit(1)

    print("Done.")


if __name__ == "__main__":
    main()
